stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -0,0 +1,127 @@
+    } catch (const cl::Error& e) {
+      check_opencl_error("cholesky_decompose", e);
+    }
+  } else {
+    L = stan::math::cholesky_decompose(A, block / divider);
+  }
+  return L;
+}
+}  // namespace internal
+/**
+ * Return the lower-triangular Cholesky factor (i.e., matrix
+ * square root) of the specified square, symmetric matrix.
+ * The return value \f$L\f$ will be a lower-traingular matrix such that the
+ * original matrix \f$A\f$ is given by
+ * <p>\f$A = L \times L^T\f$.
+ * The Cholesky decomposition is computed on the GPU. This algorithm is
+ * recursive, where The parameters <code>block</code>, <code>divider</code>, and
+ *  <code>min_block</code> act as tuning parameters for the recursive step of
+ *  the GPU based Cholesky decompostion. The matrix is subset by the
+ *  <code>block</code> size, and if the <code>block</code> size is less than
+ * <code>min_block</code> then the cholesky decomposition on the GPU is computed
+ * using that submatrix. If <code>block</code> is greater than
+ * <code>block_size</code> then <code>cholesky_decompose</code> is run again
+ * with <code>block</code> equal to <code>block/divider</code>. Once the
+ * Cholesky Decomposition is computed, the full matrix cholesky is created
+ * by propogating the cholesky forward as given in the reference report below.
+ *
+ * For a full guide to how this works
+ * see the Cholesy decompostion chapter in the  reference report
+ * <a href="https://goo.gl/6kWkJ5"> here</a>.
+ * @param A Symmetric matrix on the GPU.
+ * @param block Size of the block used to compute the cholesky decomposition.
+ * @param divider Proportion to divide the submatrix by at each recursive step.
+ * @param min_block The amount that block is checked against to decide
+ *  whether to continue the recursion or to perform the cholesky.
+ * @return Square root of matrix on the GPU.
+ * @throw std::domain_error if m is not
+ *  positive definite (if m has more than 0 elements)
+ */
+inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int block,
+                                     const int divider, const int min_block) {
+  auto offset = 0;
+  // NOTE: The code in this section follows the naming conventions
+  // in the report linked in the docs.
+  matrix_gpu A_11(block, block);
+  // Repeats the blocked cholesky decomposition until the size of the remaining
+  // submatrix is smaller or equal to the block size
+  while ((offset + block) < (A.rows())) {
+    auto block_subset = A.rows() - offset - block;
+    matrix_gpu A_21(block_subset, block);
+    matrix_gpu A_22(block_subset, block_subset);
+    // Copies a block of the input A into A_11
+    A_11.sub_block(A, offset, offset, 0, 0, block, block);
+    // The following function either calls the
+    // blocked cholesky recursively for the submatrix A_11
+    // or calls the kernel  directly if the size of the block is small enough
+    matrix_gpu L_11 = stan::math::internal::cholesky_decompose_recursion(
```

Comment:
It's sort of weird that you might have multiple blocks in A that are "too big" but you only here decrease the block size implicitly with this recursive call. So it seems like you might see execution like this:
A = 4000, 4000; block = 400, min_block = 100; divider =2

```
1. cdecompose(A, 400)
2. for each block of size 400 in A, 
    a. cdecompose_recursion(A_block, 400)
    b. which then calls cdecompose(A_block, 200)
    c. for each block of size 200 in A_block, calls cdecompose_recursion(A_block_block, 200)
         1. cdecompose(A_block_block, 100)...
```

I think this will probably cover all of the blocks correctly (need tests for different block sizes etc, see other comment), but I think the code would seem simpler to me if we first found the correct block size and then executed something on each of the blocks. I had some trouble tracing through these mutually recursive functions, granted I asked for some code [DRYing](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) up and I think this a bit easier to grok than before but maybe we can do better... I'm trying out some patterns myself here but want to get the review back to you all before I finish that up.

Reply:
Ah okay! Thanks :) Will hold off if this makes sense to you.

Reply:
If Rok has a better way to do this then I'm all for it, but here I just used `floor(m.rows() / 2)` as the starting block size in the primitive code for cholesky